admin-tools/lib/src/main/scala/com/gu/mediaservice/ImagesBatchProjection.scala

```diff
@@ -4,49 +4,60 @@ import java.net.URL
 
 import com.gu.mediaservice.model.Image
 
-import scala.concurrent.duration.Duration
+import scala.concurrent.duration.{DAYS, Duration, HOURS}
 import scala.concurrent.{Await, ExecutionContext, Future}
 
 class ImagesBatchProjection(apiKey: String, timeout: Duration, gridClient: GridClient, maxSize: Int) {
 
   private implicit val ThrottledExecutionContext = ExecutionContext.fromExecutor(java.util.concurrent.Executors.newFixedThreadPool(5))
 
-  def validApiKey(projectionEndpoint: String) = {
+  // TODO I don't like this - checking the API key by making a request to a non-existent endpoint is ... ewww.
+  def validateApiKey(projectionEndpoint: String): Unit = {
     val projectionUrl = new URL(s"$projectionEndpoint/not-exists")
-    val statusCode = gridClient.makeGetRequestSync(projectionUrl, apiKey).statusCode
-    statusCode != 401 && statusCode != 403
+    for {
+      _ <- gridClient.makeGetRequestAsync(
+        projectionUrl,
+        apiKey,
+        None,
+        {res:ResponseWrapper => None},
+        {res:ResponseWrapper => None}
+      )
+    } yield Unit
   }
 
   def getImagesProjection(mediaIds: List[String], projectionEndpoint: String,
                           InputIdsStore: InputIdsStore): List[Either[Image, String]] = {
     val apiCalls = mediaIds.map { id =>
       val projectionUrl = new URL(s"$projectionEndpoint/$id")
-      val responseFuture: Future[ResponseWrapper] = gridClient.makeGetRequestAsync(projectionUrl, apiKey)
-      val notFoundOrImage: Future[Option[Either[Image, String]]] = responseFuture.map { response =>
-        response.statusCode match {
-          case 200 if response.bodyAsString.size > maxSize => {
+      class FailedCallException extends Exception
+      gridClient.makeGetRequestAsync[Either[Image, String]](
+        projectionUrl,
+        apiKey,
+        None,
+        { response: ResponseWrapper =>
+          if (response.bodyAsString.size > maxSize) {
             InputIdsStore.setStateToTooBig(id, response.bodyAsString.size)
             None
+          } else {
+            Some(Left(response.body.as[Image]))
           }
-          case 200 => {
-            val img = response.body.as[Image]
-            Some(Left(img))
-          }
-          case 404 => {
+        },
+        { _: ResponseWrapper => {
             InputIdsStore.updateStateToNotFoundImage(id)
             Some(Right(id))
           }
-          case _ if (isAKnownError(response)) => {
+        },
+        Some({ res: ResponseWrapper => {
+          if (isAKnownError(res)) {
             InputIdsStore.setStateToKnownError(id)
-            None
-          }
-          case _ => {
+            new FailedCallException()
```

Comment:
Does this want to take the id for better logging?

Reply:
Not needed, it is caught locally.